openedx/openedx-platform
# Store content object collections in search index [FC-0062]

## Description

This PR ensures that when a content object is added/removed from a Collection, its search index document is updated to reflect this change. The `collection.key` values for each collection are stored in a new `collections` list.

To support this, we added a new event called `CONTENT_OBJECT_ASSOCIATIONS_CHANGED`, and trigger it when an object's collections change. This event is also now triggered when an object's tags change. The previous `CONTENT_OBJECT_TAGS_CHANGED` event is still emitted for tag changes, but is now deprecated.

Useful information to include:

- Which edX user roles will this change impact? Course Authors who have enabled Content Libraries V2.

## Supporting information

Related Tickets:
 - https://github.com/openedx/modular-learning/issues/230
 - Private-ref: [FAL-3787](https://tasks.opencraft.com/browse/FAL-3787)

Depends on / blocked by:
 - https://github.com/openedx/edx-platform/pull/35321 -- see https://github.com/open-craft/edx-platform/pull/680 for code review.

## Testing instructions

* Setup tutor with this PR and https://github.com/open-craft/tutor-contrib-meilisearch plugin, plus ensure you are running with the latest versions of openedx-learning + openedx-events.
* Run migrations: `tutor dev run cms python manage.py migrate`
* In the content-authoring MFE, create a library, e.g. `lib:SampleTaxonomyOrg1:AL1`
* Add some blocks to the library, and note their block keys.
  Hint: you can retrieve the block keys using the REST API, e.g. http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/blocks/
* Run `tutor dev run cms ./manage.py cms reindex_studio --experimental` to add the new `collections` field.
* Run below snippet to add some collections -- reindexing should happen automatically.
```python
# tutor dev run cms python manage.py cms shell
from opaque_keys.edx.keys import UsageKey
from opaque_keys.edx.locator import LibraryLocatorV2
from openedx.core.djangoapps.content_libraries import api                                                                                                
lib_key_str = "lib:<your lib key>"
block_key_strs = [
    "lb:<your block key",
    "lb:<your block key",
    # ....
]

library_key = LibraryLocatorV2.from_string(lib_key_str)
block_keys = [
    UsageKey.from_string(key_str) for key_str in block_key_strs
]

api.create_library_collection(library_key, "FAL-3787", title="Collection FAL-3787")
api.update_library_collection_components(library_key, "FAL-3787", usage_keys=block_keys)
# <Collection> (lp:1 FAL-3787:Collection FAL-3787) 
```

* Open http://meilisearch.local.edly.io:7700/ 
  Your api key can be found with `tutor config printvalue MEILISEARCH_API_KEY`
* Search for `FAL-3787` and verify that each of your blocks has `FAL-3787` showing in its `collections` list.

## Deadline

Before Sumac gets cut in October.
